### ISSUE TITLE
Updating from 3.2.0 to 3.2.1 we got 336 tests failing

### ISSUE THREAD

**Issue by laurentkempe** [nunit/nunit#1438]
We have some tests for our ViewModel layers that started to fail when we updated NUnit from 3.2.0 to 3.2.1.

Those tests all fails in a productive code line which does
`DispatcherHelper.CheckBeginInvokeOnUI(() => BindingOperations.EnableCollectionSynchronization(items, _itemsLock));`
DispatcherHelper is a helper from MVVMLight framework which makes life easier when working with WPF Dispatcher.
As it we are working with WPF we have some of our tests which requires STA and all have been migrated to use the new `[Apartment(ApartmentState.STA)]` those which wasn't needing it does not have it.

I have seen in the release notes of 3.2.1 that there were a change linked to STA, #1208 but cannot really make the link at the moment and we are using STA on the method level only

**Comment by rprouse** [comment]
Is there an exception thrown or other errors that might help us diagnose the problem?

Have you tried debugging the tests?

**Comment by rprouse** [comment]
@laurentkempe did you make any progress on determining the root cause of this issue? Is the dispatch helper not invoking on the UI thread? What errors/exceptions are being thrown?

**Comment by jtu100** [comment]
We are also now getting failing WPF tests with 3.2.1.

This may be a red herring but it appears that WPF's `CollectionViewSource.GetDefaultView(..)` seems to be returning null in 3.2.1 but in 3.2.0 it was working correctly.

**Comment by jtu100** [comment]
More investigation comparing two versoins of nunit in WPF:

When trying to access the data binding engine in 3.2.0 it's happening on threads like so:
>> Thread Id: 24, Name: {no name}, STA. Dispatcher: no

But each test is running with a _different _thread Id.

On 3.2.1 I see this:
>>Thread Id: 10, Name: "Worker#NP_STA", STA. Dispatcher: yes

But the _same _ thread Id is running through all my tests (this concurs with the reported change to Nunit).

The main different here is the WPF Dispatcher has been created for the thread on 3.2.1(broken tests) but not for 3.2.0 (working tests). And obviously different threads for each test.

Still unsure of the specific cause but my guess is reusing the same thread across different tests is a problem for WPF.

**Comment by CharliePoole** [comment]
One reason among others that caused us to return to the 3.0 model of using the same thread is that the user can always override this behavior and require a separate thread. See RequiresThreadAttribute.

**Comment by jtu100** [comment]
`RequiresThreadAttribute `seems to be working for me (although using it at the assembly-level does not), thanks.

As an aside, I think this API could be improved - I'd found the `SingleThreadedAttribute ` on the 3.2.1 release notes, and had searched for its opposite without success. Perhaps a single attribute `ThreadingMode ` or similar would be more discoverable? On top of that would prevent you doing, what I assume to be a mutex, `SingleThreadedAttribute `+ `RequiresThreadAttribute `at the same time?

**Comment by CharliePoole** [comment]
RequiresThreadAttribute applies only to the item it's on so it makes no
sense at assembly level. We shouldn't allow it. It's actually an attribute
I have tried to get rid of but people keep needing it. :-)

The general philosophy of NUnit is that tests are independent and therefore
can be run in any way we like. Consequently, attributes to control things
like threading have only been added to meet specific needs, at different
times. In this approach, any "systematisation" of the API will always come
later.

SingleThreaded attribute is the first time I can remember where we created
an attribute that does no more than enforce our default behavior. But it
seemed to make sense because

1. That default could change.

2. It allows us to give an error when's conflicting attributes appear.

In the 3.2.1 release a RequiresThread attribute is ignored silently if it
appears on a method of a single-threaded fixture. In the latest source it
gives an error.
